File: src/Jetpac.Core/Flags.fs
namespace Jetpac.Core

/// The Z80 flag register, ported from specbolt (z80/common/include/z80/common/Flags.hpp).
/// Stored as an int; 8-bit operators mask to 0xFF. A struct so ALU operations
/// allocate nothing (critical for emulator throughput).
[<Struct>]
type Flags =
  val Value: int
  new(value: int) = { Value = value &&& 0xFF }

  member this.ToU8() = this.Value

  static member (&&&) (lhs: Flags, rhs: Flags) = Flags(lhs.Value &&& rhs.Value)
  static member (|||) (lhs: Flags, rhs: Flags) = Flags(lhs.Value ||| rhs.Value)
  static member (^^^) (lhs: Flags, rhs: Flags) = Flags(lhs.Value ^^^ rhs.Value)
  static member (~~~) (f: Flags) = Flags(~~~f.Value &&& 0xFF)

  static member Carry() = Flags(0x01)
  static member Subtract() = Flags(0x02)
  static member Parity() = Flags(0x04)
  static member Overflow() = Flags(0x04)
  static member Flag3() = Flags(0x08)
  static member HalfCarry() = Flags(0x10)
  static member Flag5() = Flags(0x20)
  static member Zero() = Flags(0x40)
  static member Sign() = Flags(0x80)

  member this.carry = this.Value &&& 0x01 <> 0
  member this.subtract = this.Value &&& 0x02 <> 0
  member this.parity = this.Value &&& 0x04 <> 0
  member this.overflow = this.Value &&& 0x04 <> 0
  member this.half_carry = this.Value &&& 0x10 <> 0
  member this.zero = this.Value &&& 0x40 <> 0
  member this.sign = this.Value &&& 0x80 <> 0

  override this.ToString() = sprintf "Flags(0x%02x)" this.Value

File: src/JetpacFR.Web/webroot/Jetpac.Core/Flags.js

import { printf, toText } from "../fable_modules/fable-library-js.5.13.0/String.js";
import { Record } from "../fable_modules/fable-library-js.5.13.0/Types.js";
import { class_type } from "../fable_modules/fable-library-js.5.13.0/Reflection.js";

/**
 * The Z80 flag register, ported from specbolt (z80/common/include/z80/common/Flags.hpp).
 * Stored as an int; 8-bit operators mask to 0xFF. A struct so ALU operations
 * allocate nothing (critical for emulator throughput).
 */
export class Flags extends Record {
    constructor(Value) {
        super();
        this.Value = (Value | 0);
    }
    toString() {
        const this$ = this;
        return toText(printf("Flags(0x%02x)"))(this$.Value);
    }
}

export function Flags_$reflection() {
    return class_type("Jetpac.Core.Flags", undefined, Flags, class_type("System.ValueType"));
}

export function Flags_$ctor_Z524259A4(value) {
    return new Flags(value & 255);
}

export function Flags__ToU8(this$) {
    return this$.Value | 0;
}

export function Flags_op_BitwiseAnd_Z51621B00(lhs, rhs) {
    return Flags_$ctor_Z524259A4(lhs.Value & rhs.Value);
}

export function Flags_op_BitwiseOr_Z51621B00(lhs, rhs) {
    return Flags_$ctor_Z524259A4(lhs.Value | rhs.Value);
}

export function Flags_op_ExclusiveOr_Z51621B00(lhs, rhs) {
    return Flags_$ctor_Z524259A4(lhs.Value ^ rhs.Value);
}

export function Flags_op_LogicalNot_Z7353D318(f) {
    return Flags_$ctor_Z524259A4(~f.Value & 255);
}

export function Flags_Carry() {
    return Flags_$ctor_Z524259A4(1);
}

export function Flags_Subtract() {
    return Flags_$ctor_Z524259A4(2);
}

export function Flags_Parity() {
    return Flags_$ctor_Z524259A4(4);
}

export function Flags_Overflow() {
    return Flags_$ctor_Z524259A4(4);
}

export function Flags_Flag3() {
    return Flags_$ctor_Z524259A4(8);
}

export function Flags_HalfCarry() {
    return Flags_$ctor_Z524259A4(16);
}

export function Flags_Flag5() {
    return Flags_$ctor_Z524259A4(32);
}

export function Flags_Zero() {
    return Flags_$ctor_Z524259A4(64);
}

export function Flags_Sign() {
    return Flags_$ctor_Z524259A4(128);
}

export function Flags__get_carry(this$) {
    return (this$.Value & 1) !== 0;
}

export function Flags__get_subtract(this$) {
    return (this$.Value & 2) !== 0;
}

export function Flags__get_parity(this$) {
    return (this$.Value & 4) !== 0;
}

export function Flags__get_overflow(this$) {
    return (this$.Value & 4) !== 0;
}

export function Flags__get_half_carry(this$) {
    return (this$.Value & 16) !== 0;
}

export function Flags__get_zero(this$) {
    return (this$.Value & 64) !== 0;
}

export function Flags__get_sign(this$) {
    return (this$.Value & 128) !== 0;
}

